Feature #3065 benchmarking ensemble stat

bikegeek opened the pull request:
## Expected Differences ##

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[Yes]**</br>
If **yes**, please describe:</br>

- Benchmarking using the CTRACK tool that runs when the --enable-profiler option is included in the configuration step of a MET build.

Instrumented code:
- src/basic/vx_util/main.cc  
- src/tools/core/ensemble_stat/ensemble_stat.cc
- src/tools/core/ensemble_stat/ensemble_stat_conf.cc

Configuration:

   benchmark.yaml:

-    for exercising MET code via command line options
-    for exercising MET code via METplus use case(s) (i.e. wrapper code)

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[No]**</br>


## Pull Request Testing ##

- [ ] Describe testing already performed for these changes:</br>

    on host 'seneca':

    cloned MET code to:
       /d1/projects/Benchmark_EnsembleStat

   built code without and with --enable-profiler to verify the option is working as expected
   ran profiling/benchmarking tool, CTRACK on ensemble stat code via:
      -  MET command line commands:
       -   use the benchmark.yaml config file set up for running MET command
       -   use the setup.bash script to set the appropriate environment variables
       -   use the envs_for_met.bash  for setting the environment variables needed to run MET command line command
       -  METplus use case
       -   set the settings for METplus in the benchmark.yaml config file


- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

      - verify that instructions under the Contributor's Guide are correct and make sense
      - only focus on testing MET command line tools
        use the benchmark.yaml config file
        use the setup.bash script to set the appropriate environment variables
        use the envs_for_met.bash  for setting the environment variables needed to run MET command line command
        verify that results are in the output directory specified in the benchmark.yaml config file

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes ]**

- [x] Do these changes include sufficient testing updates? **[No]**

- [x] Will this PR result in changes to the MET test suite? **[No]**</br>


- [x] Will this PR result in changes to existing METplus Use Cases? **[No]**</br>
If **yes**, create a new **Update Truth** [METplus issue](https://github.com/dtcenter/METplus/issues/new/choose) to describe them.

- [x] Do these changes introduce new SonarQube findings? **[Yes]**</br>
If **yes**, please describe:
    may result in lower code coverage percentage for new code

- [x] Please complete this pull request review by **for RC1 release**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR definition above.
- [ ] Ensure the PR title matches the feature or bugfix branch name.
- [ ] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the version that will include these changes
Select: **METplus-X.Y Support** project for bugfix releases or **MET-X.Y Development** project for the next coordinated release
- [ ] After submitting the PR, select the :gear: icon in the **Development** section of the right hand sidebar. Search for the issue that this PR will close and select it, if it is not already selected.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

georgemccabe committed:
use custom GitHub Action to trigger METplus use cases

jprestop committed:
Updating values

JohnHalleyGotway committed:
Bugfix #3020 main_v12.0 grid_stat_seeps (#3022)

* Per #3020, add missing GridStatNcOutInfo::do_seeps flag and use it to determine if SEEPS information should be written to the Grid-Stat NetCDF matched pairs output file.

* Unrelated to #3020, fix broken NetCDF cf-conventions links in the User's Guide.

* Per #3020, no real changes. Just whitespace

jprestop committed:
Update to reflect usage of oneAPI compilers

jprestop committed:
Updating file to reflect usage of oneAPI compilers

JohnHalleyGotway committed:
Hotfix to the main_v12.0 branch after PR #3022 fixed a SEEPS bug. The GridStatConfig_SEEPS config file needs to be updated with nc_pairs_flag.seeps = TRUE in order for the same output to be produced by the unit tests.

jprestop committed:
Adding In Memoriam

JohnHalleyGotway committed:
Feature #3032 main_v12.0 docs data type (#3040)

* Per #3032, add data type column to all of the output tables

* Per #3032, remove the first row from each output table since its info is repeated from the table name. Additional changes for consistency and accuracy in column names.

* Update docs/Users_Guide/gsi-tools.rst

Co-authored-by: Julie Prestopnik <jpresto@ucar.edu>

---------

Co-authored-by: Julie Prestopnik <jpresto@ucar.edu>

JohnHalleyGotway committed:
Making a superficial change in the main_v12.0 branch to trigger GHA to create and push an updated test output image.

JohnHalleyGotway committed:
Feature #3033 v12.0.0 (#3042)

* Per #3033, update version info, consolidate release notes, and add upgrade instructions.

* Per #3033, remove all instances of 'Bugfix: ' from the release notes since it's redundant with the dropdown name

* Per #3030, based on request from Randy Pierce, also add MTD header columns to met_header_columns_v12.0.txt to make it easier to parse the output from MET.

* Per #3033, fix typo and correct alignment in table

jprestop committed:
Update install_met_env.acorn

Removing reference to beta version

jprestop committed:
Update install_met_env.cactus

Remove references to beta version

jprestop committed:
Update install_met_env.cactus

Update paths for eckit and atlas

jprestop committed:
Update install_met_env.wcoss2

Remove beta references

bikegeek committed:
initial YAML config file used to perform benchmarking on METplus use cases

bikegeek committed:
initial version of code used to run use case on MET that has CTRACK macro enables

JohnHalleyGotway committed:
Fix typo, missing one * to make SciPy bold in appendixF.rst

bikegeek committed:
Add support to run the use case multiple times, based on the num_runs setting in the yaml config file

bikegeek committed:
Add support for determining the path where the YAML file is located

bikegeek committed:
Allow a list of wrapper config files, create subdirectories named after use case config files, add more assertion checks

bikegeek committed:
Add list to wrapper config file

georgemccabe committed:
Per #3051, update unit tests so that installed files are found relative to MET_BASE (<install_loc>/share/met) and other files that are only in the MET repo are found relative to MET_TEST_BASE (MET/internal/test_unit). Also remove MET_BUILD_BASE env var (#3052)

bikegeek committed:
Removed incorrectly placed asserts and fixed incorrect comment

bikegeek committed:
Fixed assertion that checks output file size

bikegeek committed:
Remove assert in check, checking for a single config file when a list is expected.

bikegeek committed:
clean up comments, mv summary_output.txt and detail_output.txt files after finishing, format the information file

bikegeek committed:
Removed option to save output as csv, tabular or both.  Both formats are saved.

bikegeek committed:
Fixed name of config file

bikegeek committed:
Fix the command to get the directory where the script has been invoked

bikegeek committed:
Incorrectly passing in the use case path instead of name into generate_info()

bikegeek committed:
copy rather than move the summary and detail files generated by CTRACK

bikegeek committed:
Merge branch 'main_v12.0' of https://github.com/dtcenter/MET into benchmarking

bikegeek committed:
Modified to write summary and detail information to files

bikegeek committed:
using CTRACK macro

bikegeek committed:
using CTRACK macro to benchmark code

bikegeek committed:
Use CTRACK macro to perform benchmarking

bikegeek committed:
Updated to address seg fault when running

bikegeek committed:
Fixes to address seg fault when running the tool

bikegeek committed:
Create README.md

first version

bikegeek committed:
Added call to ctrack to write out results in the "main" function met_main()

bikegeek committed:
ctrack updated to write summary and details to files

bikegeek committed:
Merge branch 'benchmarking' of https://github.com/dtcenter/MET into benchmarking

bikegeek committed:
Update README.md

added reminder to add 
    ctrack::result_print();
to ensure the summary and detail files are written

bikegeek committed:
CTRACK include file

bikegeek committed:
Merge branch 'benchmarking' of https://github.com/dtcenter/MET into benchmarking

bikegeek committed:
Removed hard-coded paths and replaced output_path with benchmark_output_path to separate output from METplus Use case output

bikegeek committed:
Clean up parsing yaml config, read the benchmark_output_path in YAML instead of output_path

bikegeek committed:
Write out ctrack information to files

bikegeek committed:
Write out ctrack results to files

bikegeek committed:
Update README.md

Finished description of settings in YAML config

bikegeek committed:
Update README.md

update to instructions- yaml config file doesn't need to be specified at the command line

bikegeek committed:
Merge branch 'develop' of https://github.com/dtcenter/MET into benchmarking

bikegeek committed:
Adding support to read in updates to YAML file to allow running MET command

bikegeek committed:
Merge branch 'benchmarking' of https://github.com/dtcenter/MET into benchmarking

bikegeek committed:
Replace latest_benchmark.yaml with benchmark.yaml

bikegeek committed:
Merge branch 'benchmarking' of https://github.com/dtcenter/MET into benchmarking

bikegeek committed:
Added config support for specifying running MET directly

bikegeek committed:
Removed remaining reference to latest_benchmark.yaml

bikegeek committed:
Added setting for specifying a subdirectory to save output when running via MET

bikegeek committed:
Refactoring the code for running usecases and adding support for running MET directly

bikegeek committed:
fixed indentation error

bikegeek committed:
Merge branch 'develop' of https://github.com/dtcenter/MET into feature_3065_benchmarking_ensemble_stat

bikegeek committed:
Remove CTRACK macro and include

bikegeek committed:
Removing ctrack header to move to vx_util

bikegeek committed:
Remove CTRACK macro and include

bikegeek committed:
Removed and placed under the vx_util directory

bikegeek committed:
Added ctrack include, use macro for pre- and post-processing functions

bikegeek committed:
Add the CTRACK benchmarking tool

bikegeek committed:
Update checking and saving information to be generic and support running via METplus wrappers or MET directly

bikegeek committed:
Merge branch 'develop' of https://github.com/dtcenter/MET into feature_3065_benchmarking_ensemble_stat

bikegeek committed:
Modify check_settings to check settings for running use cases or running MET directly

bikegeek committed:
Instrument with CTRACK macro for benchmarking

bikegeek committed:
Support setting a description (optional) which will appear in the information file

bikegeek committed:
Include instructions to save the benchmarking summary and detail information to files

bikegeek committed:
Fix setting name from met_cmd to met_command to match what is in the YAML file

bikegeek committed:
Pass MET command as detail into the info.txt file

bikegeek committed:
Convert full_subdir_path to string

bikegeek committed:
Pass in the full subdir path to extract_summary_info and extract_detail_info to get correct behavior

bikegeek committed:
Make the MET final output subdirectory (where consolidated files go) if it doesn't exist

bikegeek committed:
Remove call to printing the ctrack report to file

bikegeek committed:
Fix input to generate_info: replace "met_command" with met_command

bikegeek committed:
Merge branch 'feature_3065_benchmarking_ensemble_stat' of https://github.com/dtcenter/MET into feature_3065_benchmarking_ensemble_stat

bikegeek committed:
Add the METPLUS_ environment information to the information text file when running the MET command directly

bikegeek committed:
set additional text to None if there is no additional information (i.e. use case run)

bikegeek committed:
Merge branch 'feature_3065_benchmarking_ensemble_stat' of https://github.com/dtcenter/MET into feature_3065_benchmarking_ensemble_stat

bikegeek committed:
Fixed order for generate_info args inside run_met_cli

bikegeek committed:
Merge branch 'feature_3065_benchmarking_ensemble_stat' of https://github.com/dtcenter/MET into feature_3065_benchmarking_ensemble_stat

bikegeek committed:
Add the values of the METplus ENV vars used in run_met_cli

bikegeek committed:
Merge branch 'feature_3065_benchmarking_ensemble_stat' of https://github.com/dtcenter/MET into feature_3065_benchmarking_ensemble_stat

bikegeek committed:
Merge branch 'develop' of https://github.com/dtcenter/MET into feature_3065_benchmarking_ensemble_stat

bikegeek committed:
Merge branch 'develop' of https://github.com/dtcenter/MET into feature_3065_benchmarking_ensemble_stat

bikegeek committed:
Merge branch 'develop' of https://github.com/dtcenter/MET into feature_3065_benchmarking_ensemble_stat

bikegeek committed:
Merge branch 'feature_3065_benchmarking_ensemble_stat' of https://github.com/dtcenter/MET into feature_3065_benchmarking_ensemble_stat

bikegeek committed:
Removed conflict

bikegeek committed:
Updates added for running MET commands

bikegeek committed:
Adding constants for time units.  This will be used when calculating the mean metric when multiple runs are performed for stress-testing.

bikegeek committed:
Latest updates to first iteration of the benchmarking tool

bikegeek committed:
Clean up of format and instructions

bikegeek committed:
Clean up some grammar

bikegeek committed:
Fix some formatting of code snippets

bikegeek committed:
Code snippets cannot have any italics or bold text

bikegeek committed:
Still have issues with formatting code snippets

bikegeek committed:
New lines not being kept in code

bikegeek committed:
more experimentation with formatting code snippet

bikegeek committed:
Remove special characters around code snippet

bikegeek committed:
more experimenting with additional newlines to format code

bikegeek committed:
Removed all code block syntax and use only newlines and indentation to format code blocks

bikegeek committed:
Re-arrange instructions for printing reports

bikegeek committed:
Update ensemble_stat.cc

Add the code to met_main() to save the CTRACK metrics

bikegeek committed:
Merge branch 'develop' of https://github.com/dtcenter/MET into feature_3065_benchmarking_ensemble_stat

bikegeek committed:
Added ifdef WITH_PROFILER for turning on CTRACK at compile time

bikegeek committed:
Instrument code for benchmarking with CTRACK

bikegeek committed:
Adding documentation for running the code profiler used in benchmarking

bikegeek committed:
updated for use in the Contributor's Guide

bikegeek committed:
Moved content to the Contributor's Guide

bikegeek committed:
Replaced formatting and cleaned up content

bikegeek committed:
Clean up some formatting used by the README.md but incompatible with RST

bikegeek committed:
Added missing information for environments for running MET commands, cleaned up formatting

bikegeek committed:
Rearrange some text for clarity

bikegeek committed:
Fixed the introduced conflicts from earlier pull

bikegeek committed:
Reorganized to reduce clutter and redundant content.

bikegeek committed:
Added clarification to Compile MET code section

bikegeek committed:
Replace the 'r' in the regex compile that was somehow removed

bikegeek committed:
Merge branch 'feature_3065_benchmarking_ensemble_stat' of https://github.com/dtcenter/MET into feature_3065_benchmarking_ensemble_stat

bikegeek committed:
Open the summary_out.txt and detail_out.txt files with append to prevent clobbering metrics when switching from profiling functions from different files.

bikegeek committed:
Merge branch 'feature_3065_benchmarking_ensemble_stat' of https://github.com/dtcenter/MET into feature_3065_benchmarking_ensemble_stat

bikegeek committed:
Merge branch 'develop' of https://github.com/dtcenter/MET into feature_3065_benchmarking_ensemble_stat

bikegeek committed:
Fixed merging of summary and detail data to ensure the matching file/function/lines are merged.

bikegeek committed:
Renamed new index to profiling_index for clarity

bikegeek committed:
Updates with information about the num_runs setting and where the summary and detail files are moved.

JohnHalleyGotway reviewed (changes requested):
I began reviewing this today and see several files that do not belong in this PR. I'm pretty confident you didn't intend to add them.

Recommend that you make the following changes:
- [x] Remove `README.install_release` since documentation should live in the User's Guide and/or Contributor's Guide instead.
- [x] Remove `benchmark.py` from the top-level directory. That script actually lives in `internal/scripts/benchmark/benchmark.py` but this is an earlier draft at the top-level.
- [x] Remove `detail_output.txt` and `summary_output.txt` from the top-level directory. These are likely output files from testing the benchmarking script that you didn't intent to actually add to the repo.
- [x] In `internal/scripts/installation/config/` revert diffs in `install_met_env.cactus` and `install_met_env.wcoss2`. These are clearly merge conflicts that don't apply to your PR.
- [x] Remove `scripts/detail_output.txt` and `script/summary_output.txt` which are likely output from testing and shouldn't be added to the repo.
- [x] Undo changes in `grid_stat.cc`, `grid_stat_conf_info.cc`, `point_stat.cc`, and `point_stat_conf_info.cc` since you're just adding spurious whitespace.

JohnHalleyGotway reviewed (changes requested):
Please see comments above.

bikegeek committed:
Removing, the correct version is located in the MET/internal/scripts/benchmark directory

bikegeek committed:
Inadvertently added but not relevant information for this commit

bikegeek committed:
No longer necessary, this information is in the Contributor's Guide and developers automatically refer to the ReadTheDocs documentation.

bikegeek commented:
@JohnHalleyGotway I removed the erroneous files and also removed the README.md in the MET/internal/scripts/benchmark directory.  The PR is now active.

bikegeek committed:
Delete summary_output.txt

should not be in source code

bikegeek committed:
Delete detail_output.txt

output should not be included with the source code

bikegeek committed:
Delete internal/scripts/benchmark/summary_output.txt

output should not be included with source code

bikegeek committed:
Delete internal/scripts/benchmark/detail_output.txt

output should not be included in the source code

bikegeek committed:
Update system.conf

remove hard-coded values

bikegeek committed:
Delete internal/scripts/benchmark/constants.py

not used for this code

bikegeek committed:
Update benchmark.yaml

num_of_runs should match instructions in the Contributor's Guide

bikegeek committed:
Delete internal/scripts/benchmark/ctrack.hpp

Correct location is in the src/basic/vx_util directory

bikegeek committed:
Update example_MET.yaml

remove hard-coded paths and set the num_of_runs to 1 to match instructions in documentation

bikegeek committed:
Update example_usecase.yaml

remove hard-coded values and set the num_of_runs to 1 to be consistent with instructions

bikegeek committed:
Merge branch 'develop' of https://github.com/dtcenter/MET into feature_3065_benchmarking_ensemble_stat

bikegeek committed:
Output files should not be included with source code

bikegeek committed:
update to version in the develop branch

bikegeek committed:
Revert to version in develop, undoing whitespace changes

bikegeek committed:
Revert to version in develop branch, eliminating whitespace changes

bikegeek commented:
I also updated the yaml config files to remove hard-coded paths and removed the unused constants.py file (to be used for stress-testing support that isn't available in this code)

davidalbo commented:
Assuming this is a first example of profiling to come, I'm wondering about putting the profiling in so many places in ensemble_stat.  Not that I'm against that but is there a good reason for it?  In other words, what is the intent of profiling this particular application, and does the way it is set up help accomplish that?

bikegeek commented:
I was taking a guess at what functions might be of interest to instrument
since I'm not familiar with the code.  If you don't think all these
functions should be instrumented, feel free to undo or let me know which
ones to undo.

On Thu, May 1, 2025 at 11:44 AM davidalbo ***@***.***> wrote:

> *davidalbo* left a comment (dtcenter/MET#3137)
> <https://github.com/dtcenter/MET/pull/3137#issuecomment-2845345418>
>
> Assuming this is a first example of profiling to come, I'm wondering about
> putting the profiling in so many places in ensemble_stat. Not that I'm
> against that but is there a good reason for it? In other words, what is the
> intent of profiling this particular application, and does the way it is set
> up help accomplish that?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dtcenter/MET/pull/3137#issuecomment-2845345418>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4UJHUWOX3226CE2CQU4H324JMRTAVCNFSM6AAAAAB3XULGE2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQNBVGM2DKNBRHA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

davidalbo commented:
Thinking out loud, I'd describe my thinking like this:   Start at a high level, look for bottlenecks, and drill down from there, meaning put only a few ctrack profiling lines initially and see where the app is spending all it's time, then look deeper.  If you put profiling everywhere it might be harder to see where to investigate/modify.    If that makes sense, I can give it a go as far as making it look like what I think would be a good initial setup.  If that does NOT make sense,  that is also ok.  Let me know what you think.

bikegeek commented:
That sounds good.  Thanks Dave!

On Mon, May 5, 2025 at 10:15 AM davidalbo ***@***.***> wrote:

> *davidalbo* left a comment (dtcenter/MET#3137)
> <https://github.com/dtcenter/MET/pull/3137#issuecomment-2851519060>
>
> Thinking out loud, I'd describe my thinking like this: Start at a high
> level, look for bottlenecks, and drill down from there, meaning put only a
> few ctrack profiling lines initially and see where the app is spending all
> it's time, then look deeper. If you put profiling everywhere it might be
> harder to see where to investigate/modify. If that makes sense, I can give
> it a go as far as making it look like what I think would be a good initial
> setup. If that does NOT make sense, that is also ok. Let me know what you
> think.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dtcenter/MET/pull/3137#issuecomment-2851519060>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4UJHW4VOXIHECICXC2P3L246FB7AVCNFSM6AAAAAB3XULGE2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQNJRGUYTSMBWGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

davidalbo commented:
@bikegeek was there a command line option to turn on the profiling?  I'd like to test out my changes.

bikegeek commented:
Contributor's Guide:
https://metplus.readthedocs.io/projects/met/en/feature_3065_benchmarking_ensemble_stat/Contributors_Guide/code_profiling.html

Section 7.1.3.1
Step 2 Compile MET Code

On Mon, May 5, 2025 at 10:43 AM davidalbo ***@***.***> wrote:

> *davidalbo* left a comment (dtcenter/MET#3137)
> <https://github.com/dtcenter/MET/pull/3137#issuecomment-2851605768>
>
> @bikegeek <https://github.com/bikegeek> was there a command line option
> to turn on the profiling? I'd like to test out my changes.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dtcenter/MET/pull/3137#issuecomment-2851605768>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4UJHRPNWBA46VCTKF5XX3246IJ3AVCNFSM6AAAAAB3XULGE2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQNJRGYYDKNZWHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

davidalbo commented:
Hopefully the last question.  Is there a way to look at these files in a way that formats them nicely?  Just treating them as ascii gives output that is hard to read.

detail_output.txt
summary_output.txt

bikegeek commented:
The metrics files are consolidated by the Python code and stored in the
directory as indicated in the YAML config file: *benchmark_output_path
*setting,
then if you specified the filename, the final file is
<filename>_timestamp.csv, <filename>_timestamp.txt.

If no filename is specified in the filename setting, then the consolidated
files are timestamp.csv, timestamp.txt.  *You can then view the csv file
via spreadsheet.  *I prefer to view using jupyter lab and opening up into a
dataframe view.

Here is the example config file:
https://metplus.readthedocs.io/projects/met/en/feature_3065_benchmarking_ensemble_stat/Contributors_Guide/code_profiling.html

Expand the drop down:
section 7.1.3.1 under the Edit the benchmark.yaml configuration file

On Mon, May 5, 2025 at 11:14 AM davidalbo ***@***.***> wrote:

> *davidalbo* left a comment (dtcenter/MET#3137)
> <https://github.com/dtcenter/MET/pull/3137#issuecomment-2851696160>
>
> Hopefully the last question. Is there a way to look at these files in a
> way that formats them nicely? Just treating them as ascii gives output that
> is hard to read.
>
> detail_output.txt
> summary_output.txt
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dtcenter/MET/pull/3137#issuecomment-2851696160>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4UJHQJMJROWO6JEMHD3ML246L7PAVCNFSM6AAAAAB3XULGE2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQNJRGY4TMMJWGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

davidalbo commented:
I reduced the profiling to the two main function calls only.  process_n_vld and process_vx.  This shows that process_vx is where the bulk of the time is being spent, so that is where one might add more profiling to see where within that method it is getting bogged down.  This is what I was after.  If you're good with this, I'll commit and push my changes. 
<img width="1026" alt="Screenshot 2025-05-05 at 1 17 23 PM" src="https://github.com/user-attachments/assets/eb9d3ac0-1ca7-4e92-adae-e75b9a6847b4" />

bikegeek commented:
Sounds good, thanks Dave!

On Mon, May 5, 2025 at 1:20 PM davidalbo ***@***.***> wrote:

> *davidalbo* left a comment (dtcenter/MET#3137)
> <https://github.com/dtcenter/MET/pull/3137#issuecomment-2852093750>
>
> I reduced the profiling to the two main function calls only. process_n_vld
> and process_vx. This shows that process_vx is where the bulk of the time is
> being spent, so that is where one might add more profiling to see where
> within that method it is getting bogged down. This is what I was after. If
> you're good with this, I'll commit and push my changes.
> Screenshot.2025-05-05.at.1.17.23.PM.png (view on web)
> <https://github.com/user-attachments/assets/eb9d3ac0-1ca7-4e92-adae-e75b9a6847b4>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dtcenter/MET/pull/3137#issuecomment-2852093750>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4UJHQLT2PVIHQ6ODTO4632462YHAVCNFSM6AAAAAB3XULGE2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQNJSGA4TGNZVGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

davidalbo committed:
Reduced the number of profiling points to just the two major top level methods

davidalbo reviewed (dismissed):
I approve this pull request.

JohnHalleyGotway commented on docs/Contributors_Guide/code_profiling.rst:
```diff
@@ -0,0 +1,538 @@
+         The ctrack.hpp file is saved in the $HOME/MET/src/basic/vx_util directory and does not need to be modified
+         or added to any other location.  This version of ctrack.hpp has been modified to write the summary and detail
+         tables to text files. By default, CTRACK is disabled and is enabled at compilation time via the
+         **--enable-profiler flag**.
```
```suggestion
         :code:`--enable-profiler` flag.
```
The `--enable-profiler` doesn't render well in the docs. Recommend using code formatting.

JohnHalleyGotway commented on docs/Contributors_Guide/code_profiling.rst:
```diff
@@ -0,0 +1,538 @@
+  https://github.com/Compaile/ctrack
+
+This code is licensed under the MIT License:
+https://github.com/Compaile/ctrack/blob/main/LICENSE
```
```suggestion
  https://github.com/Compaile/ctrack/blob/main/LICENSE
```
Consistent indentation for these URL's.

JohnHalleyGotway commented on src/tools/core/ensemble_stat/ensemble_stat.cc:
```diff
@@ -194,6 +198,11 @@ int met_main(int argc, char *argv[]) {
 
+  // Save the CTRACK metrics 
+  #ifdef WITH_PROFILER
+  ctrack::result_print();
```
@bikegeek if `ctrack::result_print()` is already being called every time in `basic/vx_util/main.cc` don't we NOT need to call it in the application code? Won't this cause the results to be printed twice?

bikegeek replied:
If ctrack::result_print() is NOT in the application code, the benchmarking information for the application code does not get saved.

JohnHalleyGotway reviewed (commented):
@bikegeek, apologies for the long delay in getting to this. I proposed a couple of minor edits in the docs.

I tested in seneca as the `met_test` user in `/d1/projects/MET/MET_pull_requests/met-12.1.0/rc1/MET-feature_3065_benchmarking_ensemble_stat` and compiled with `--enable-profiler`.

When I run `make test`, I do see all the tests being run, including CTRACK output, and that's good. But it fails with this error at the end:
```
make[1]: *** No rule to make target 'profiler', needed by 'all'.  Stop.
make[1]: Leaving directory '/d1/projects/MET/MET_pull_requests/met-12.1.0/rc1/MET-feature_3065_benchmarking_ensemble_stat/scripts'
make: *** [Makefile:880: test] Error 2
```
I'll see if I can figure out why that's happening.

Commit:
Per #3065, fix scripts/Makefile to ignore ENABLE_PROFILER when determining which tests to run.

JohnHalleyGotway reviewed (dismissed):
OK, I just pushed a fix to the `make test` error. It's because of custom logic in `scripts/Makefile` to determine which test scripts to run. But I patched it to ignore the `--enable-profiler` configuration setting.

I approve of this PR, but do recommend that you make those documentation tweaks.

I imagine as we start using this profiling option, we may want to modify the details. @georgemccabe mentioned one idea to push the CTRACK directives in a common library function, so that adding it to a function is a one-liner instead of a three-liner.

Thanks!

bikegeek committed:
Update code_profiling.rst

No more error message during make test. Update comiler section accordingly

bikegeek committed:
Update docs/Contributors_Guide/code_profiling.rst

Co-authored-by: John Halley Gotway <johnhg@ucar.edu>

bikegeek committed:
Update docs/Contributors_Guide/code_profiling.rst

Co-authored-by: John Halley Gotway <johnhg@ucar.edu>

bikegeek commented:
@davidalbo / @JohnHalleyGotway 

I updated the documentation and just need one of you to click 'approve'

JohnHalleyGotway reviewed (approved):
I approve of these changes.

I'm sorry I missed the ping on this yesterday!
